### PR TITLE
Add Width Constraints with Annotations

### PR DESCRIPTION
**Pull request by albertchen-sifive**
This changes the `InferWidths` transform. `InferWidths` now collects extra constraints from `WidthGeqConstraintAnnotation` and runs with those extra constraints. Also makes `InferWidths` a `Transform` class instead of a `Pass` object.

This transform can be used to add width constraints that are introduced by later transforms like `WiringTransform`.

Related issue: https://github.com/freechipsproject/firrtl/issues/933

### CODE REVIEW

File: src/main/scala/firrtl/transforms/InferWidthsWithAnnos.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        case w: DefWire => Some(w.name -> w.tpe)
+        case r: DefRegister => Some(r.name -> r.tpe)
+        case n: DefNode => Some(n.name -> n.value.tpe)
+        case i: WDefInstance => Some(i.name -> i.tpe)
+        case m: DefMemory => Some(m.name -> MemPortUtils.memType(m))
+        case other => None
+      }
+      pairOpt.foreach { case (ref, tpe) =>
+        typeMap += (ReferenceTarget(circuitName, modName, Nil, ref, Nil) -> tpe)
+      }
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/InferWidthsWithAnnos.scala]: Why don't you update the typeMap directly in the stmt match?

Comment (albertchen-sifive) [reply src/main/scala/firrtl/transforms/InferWidthsWithAnnos.scala]: no real reason, `Some(ref -> tpe)` was just much less typing than `typeMap += (ReferenceTarget(circuitName, modName, Nil, ref, Nil) -> tpe)`

Comment (jackkoenig) [reply src/main/scala/firrtl/transforms/InferWidthsWithAnnos.scala]: It is a little bit DRYer this way

File: src/main/scala/firrtl/passes/InferWidths.scala [hunk 2]
```diff
@@ -360,10 +386,56 @@ object InferWidths extends Pass {
...
+      mod.ports.foreach { port =>
+        typeMap += (ReferenceTarget(circuitName, mod.name, Nil, port.name, Nil) -> port.tpe)
+      }
+      mod.foreachStmt(getDeclTypes(mod.name))
+    }
+
+    val extraConstraints = state.annotations.flatMap {
+      case anno: WidthGeqConstraintAnnotation if anno.loc.isLocal && anno.exp.isLocal  =>
+        val locType :: expType :: Nil = Seq(anno.loc, anno.exp) map { target =>
+          val baseType = typeMap(target.copy(component = Seq.empty))
```
Comment (azidar) [inline src/main/scala/firrtl/passes/InferWidths.scala]: Need to check whether typeMap contains the reference, its possible that something else went wrong and they don't match up.

File: src/main/scala/firrtl/passes/InferWidths.scala [hunk 3]
```diff
@@ -360,10 +386,56 @@ object InferWidths extends Pass {
...
+        case m: DefMemory => Some(m.name -> MemPortUtils.memType(m))
+        case other => None
+      }
+      pairOpt.foreach { case (ref, tpe) =>
+        typeMap += (ReferenceTarget(circuitName, modName, Nil, ref, Nil) -> tpe)
+      }
+      stmt.foreachStmt(getDeclTypes(modName))
+    }
+
+    state.circuit.modules.foreach { mod =>
```
Comment (azidar) [inline src/main/scala/firrtl/passes/InferWidths.scala]: Maybe only walk AST after confirming that you do have annotations? That way you save an AST walk in the common case.

**Comment by azidar** [comment]
Also, please update the PR description to say that it affects the InferWidths transform, not that it adds a new transform.

**Comment by azidar** [comment]
Will commenting trigger mergify?

**Comment by jackkoenig** [comment]
lol apparently

**Pull request merged**